language/move-vm/types/src/values/values_impl.rs

```diff
@@ -1811,6 +1832,91 @@ impl Vector {
     }
 }
 
+/***************************************************************************************
+ *
+ * Runtime Type checks
+ *
+ **************************************************************************************/
+
+impl Value {
+    pub fn check_type(&self, ty: &Type) -> PartialVMResult<()> {
+        self.0.check_type(ty)
+    }
+}
+
+impl ValueImpl {
+    fn check_type(&self, ty: &Type) -> PartialVMResult<()> {
+        match (self, ty) {
+            (ValueImpl::U8(_), Type::U8)
+            | (ValueImpl::U128(_), Type::U128)
+            | (ValueImpl::U64(_), Type::U64)
+            | (ValueImpl::Bool(_), Type::Bool)
+            | (ValueImpl::Address(_), Type::Address) => Ok(()),
+            (ValueImpl::IndexedRef(r), Type::Reference(ty) | Type::MutableReference(ty)) => {
+                r.copy_value().read_ref()?.check_type(ty)
+            }
+            (ValueImpl::ContainerRef(r), Type::Reference(ty) | Type::MutableReference(ty)) => {
+                r.container().check_type(ty)
+            }
+            (ValueImpl::Container(c), ty) => c.check_type(ty),
+            (ValueImpl::U8(_), _)
+            | (ValueImpl::U128(_), _)
+            | (ValueImpl::U64(_), _)
+            | (ValueImpl::Bool(_), _)
+            | (ValueImpl::Address(_), _)
+            | (ValueImpl::IndexedRef(_), _)
+            | (ValueImpl::ContainerRef(_), _)
+            | (ValueImpl::Invalid, _) => Err(PartialVMError::new(
+                StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+            )
+            .with_message(format!(
+                "unexpected type mismatch between {:?}, expected {:?}",
+                self, ty
+            ))),
+        }
+    }
+}
+
+impl Container {
+    fn check_type(&self, ty: &Type) -> PartialVMResult<()> {
+        match (self, ty) {
+            (Self::VecAddress(_), _)
+            | (Self::VecBool(_), _)
+            | (Self::VecU128(_), _)
+            | (Self::VecU64(_), _)
+            | (Self::VecU8(_), _) => Ok(()),
```

Comment:
I don't think this is correct. We should check if the type is indeed a `vector<address> or vector<bool>, ...`

Reply:
Fixed... I thought those values to be uninteresting so didn't implement that in my first draft :(